### ISSUE TITLE
Add more C API tests

### ISSUE THREAD

**Issue by serhiy-storchaka** [python/cpython#111495]
- [x] Add tests for PyBytes C API
- [x] Add tests for PyByteArray C API
- [x] Add tests for PyFloat C API
- [x] Add tests for PyComplex C API
- [x] Add tests for PyLong C API
- [x] Add tests for PyNumber C API
- [x] Add tests for PyTuple C API
- [x] Add tests for PyList C API
- [x] Add tests for PyCodec C API
- [x] Add tests for PyFile C API

### Linked PRs
* gh-111496
* gh-111562
* gh-111586
* gh-111591
* gh-111592
* gh-111606
* gh-111607
* gh-111624
* gh-111631
* gh-111709
* gh-111731
* gh-111752
* gh-111753
* gh-111861
* gh-111996
* gh-118757
* gh-122789
* gh-122833
* gh-122851
* gh-123338
* gh-123343
* gh-123371
* gh-123375
* gh-123376
* gh-123379
* gh-124915
* gh-126030
* gh-128987
* gh-129023
* gh-129449
* gh-129477
* gh-129501

**Comment by skirpichev** [comment]
@serhiy-storchaka, initial set of C API tests for PyComplex (only for changed functions) was added in #109647

**Comment by serhiy-storchaka** [comment]
Yes, there are also some tests for PyLong and PyNumber, and maybe other parts of the C API. But we should test the full C API, or at least document what is not tested. And it is better to group tests by the prefix, so it is clear where to find tests.

**Comment by serhiy-storchaka** [comment]
If you want to work on adding tests to the part of the C API, say it here to avoid duplicating the work of other people. I finished work on PyBytes and PyByteArray and am not currently working on any of the above.

**Comment by serhiy-storchaka** [comment]
Good catch, @skirpichev. I forgot to add the actual test files.

**Comment by rawwar** [comment]
@serhiy-storchaka , I would like to contribute tests to PyTuple and PyList C API

**Comment by serhiy-storchaka** [comment]
> @serhiy-storchaka , I would like to contribute tests to PyTuple and PyList C API

Great. Look at existing PySequence tests in `Lib/test/test_capi/test_abstract.py`. Some PyTuple and PyList tests can be a copy of PySequence tests. Look also at PyDict and PySet tests.

**Comment by skirpichev** [comment]
With multiple prs, Modules/Setup.stdlib.in will produce merge conflicts.  (I already have fixed this a lot in https://github.com/python/cpython/pull/109647.)  I will create a pr to add stub files for testing modules.

**Comment by serhiy-storchaka** [comment]
> I will create a pr to add stub files for testing modules.

Good idea!

**Comment by skirpichev** [comment]
Ok, now I'll work on PyNumber_* tests.

**Comment by skirpichev** [comment]
@serhiy-storchaka, it seems missing tests for PyLong_* C API were added in #110629 or do you keep in mind some TODO's here?  For example, we could make separate tests for PyLong_From/PyLong_As functions.  Something else?

**Comment by serhiy-storchaka** [comment]
I forgot about it.

**Comment by skirpichev** [comment]
I'm working on PyTuple_* tests.

**Comment by picnixz** [comment]
I've updated the bullet point list. Out of the remaining tasks, there is only the Codec API that needs to have more tests. The PyNumber and PyFile API seem to have PRs (I didn't check whether they are up-to-date or not). Can I work on the Codec API?

**Comment by vstinner** [comment]
> The PyNumber and PyFile API seem to have PRs (I didn't check whether they are up-to-date or not).

I just merged the PyNumber PR.

> Can I work on the Codec API? (I've never added tests to the C API so I want to learn it with this)

You can go ahead.

**Comment by vstinner** [comment]
@serhiy-storchaka: I added tests on the PyFile C API with https://github.com/python/cpython/pull/129449. It was the last checkbox. Can we close this issue now?

**Comment by vstinner** [comment]
I close the issue. A lot of work has been done in this issue!

If someone wants to continue this work, I suggest to open new issues such as https://github.com/python/cpython/issues/117968.
